### ISSUE TITLE
[no-explicit-any] allow `any` in a `keyof any` expression

### ISSUE THREAD

**Issue by iliazeus** [typescript-eslint/typescript-eslint#3206]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-explicit-any": 2
  }
}
```

```TS
function getNumber<K extends keyof any, T extends { [P in K]: number }>(key: K, obj: T): number {
  return obj[key];
}
```

Currently, the expression `keyof any` is probably the only (sane) way of saying "a type of things that can be object keys". But it is disallowed under `no-explicit-any`.

The long way of `string | number | symbol` isn't portable enough because of the `symbol` part. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.18.0` |
| `@typescript-eslint/parser`        | `4.18.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.22.0` |
| `node`                             | `14.15.4` |

**Comment by iliazeus** [comment]
A more realistic example would be a decorator that can only be applied to `number` properties:

```TS
const integer = <TKey extends keyof any, TTarget extends { [K in TKey]: number }>(
  target: TTarget,
  key: TKey
) => { /* ... */ };

class Foo {
  @integer foo = 123;
  @integer bar = "adsf"; // error: 
}
```

**Comment by bradzacher** [comment]
You can use `never` instead if you must use the `keyof` syntax.
Or you could also just use the `PropertyKey` type declared as part of the TS types

```ts
const integer = <TKey extends PropertyKey , TTarget extends { [K in TKey]: number }>(
  target: TTarget,
  key: TKey
) => { /* ... */ };

class Foo {
  @integer foo = 123;
  @integer bar = "adsf"; // error: 
}
```
[repl](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=10&pc=1#code/MYewdgzgLgBAlmKBTA5kgTjAvDAPAFQGkkBPGJAD2TABMIYAFdEABwyhOLIBoZ98AhujSxK1OjADeMANqF4YPlwC6ALhhgArgFsARhhgBfAHwAKALAAoGDChCR6-vaRRuVmwGtSjrlYCU2MZSMAD0AFQwAHTRMGEhRgDcVlbAADYCEPQAYiAgUu4wAAIIyGiYAGa52DAAjABMAMxJ1kUlqAa6QtUARAJ05d0JofEYzOjqVoZWQA)
